From 04c1fb42ad3f7d1c38cfc281200ed23e6f78a366 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
Date: Wed, 11 Dec 2024 18:01:34 +0800
Subject: [PATCH 107/146] mtk: wifi: mt76: mt7996: Add mt7996_debug_mask
 support

Add mt7996_debug_mask support
Remove the following functions and change them to __mt7996_dbg()
mt76_vif_dbg()
mt76_dbg()
mt76_trace()

Usage -
The default value of debug mask is 0x1f,
which means enabling all logs except TXRX.

To check the value of debug mask:
cat /sys/module/mt7996e/parameters/mt7996_debug_mask

There are two methods to modify the value:
1. vi /etc/modules.d/mt7996e
mt7996e mt7996_debug_mask=<new value>
reboot

2. reload mt7996e module with the debug mask value
modprobe mt7996e mt7996_debug_mash=<new value>

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 mt76_connac_mcu.c |  1 -
 mt7996/Makefile   |  3 +-
 mt7996/debug.c    | 65 ++++++++++++++++++++++++++++++++
 mt7996/debug.h    | 41 ++++++++++++++++++++
 mt7996/mac.c      |  3 --
 mt7996/main.c     | 95 ++++++++++++++++++++++++++++++++++-------------
 mt7996/mcu.c      | 47 ++++++++++++++---------
 7 files changed, 206 insertions(+), 49 deletions(-)
 create mode 100644 mt7996/debug.c
 create mode 100644 mt7996/debug.h

diff --git a/mt76_connac_mcu.c b/mt76_connac_mcu.c
index dcd615ae..d33bb61f 100644
--- a/mt76_connac_mcu.c
+++ b/mt76_connac_mcu.c
@@ -427,7 +427,6 @@ void mt76_connac_mcu_sta_basic_tlv(struct mt76_dev *dev, struct sk_buff *skb,
 	}
 
 	memcpy(basic->peer_addr, link_sta->addr, ETH_ALEN);
-	pr_info("%s: link %u addr [%pM]\n", __func__, link_sta->link_id, basic->peer_addr);
 	basic->qos = link_sta->sta->wme;
 }
 EXPORT_SYMBOL_GPL(mt76_connac_mcu_sta_basic_tlv);
diff --git a/mt7996/Makefile b/mt7996/Makefile
index c9299490..d1593a31 100644
--- a/mt7996/Makefile
+++ b/mt7996/Makefile
@@ -13,4 +13,5 @@ mt7996e-y := pci.o init.o dma.o eeprom.o main.o mcu.o mac.o \
 mt7996e-$(CONFIG_DEV_COREDUMP) += coredump.o
 mt7996e-$(CONFIG_NL80211_TESTMODE) += testmode.o
 
-mt7996e-y += mtk_debugfs.o mtk_mcu.o mtk_debugfs_i.o trace.o
+mt7996e-y += mtk_debugfs.o mtk_mcu.o mtk_debugfs_i.o trace.o \
+	     debug.o
diff --git a/mt7996/debug.c b/mt7996/debug.c
new file mode 100644
index 00000000..2bcb3669
--- /dev/null
+++ b/mt7996/debug.c
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: ISC
+/*
+ * Copyright (C) 2024 MediaTek Inc.
+ */
+
+#include <linux/vmalloc.h>
+#include "debug.h"
+#include "mt7996.h"
+
+void mt7996_info(struct mt7996_dev *dev, const char *fmt, ...)
+{
+	struct va_format vaf = {
+		.fmt = fmt,
+	};
+	va_list args;
+
+	va_start(args, fmt);
+	vaf.va = &args;
+	dev_info(dev->mt76.dev, "%pV", &vaf);
+
+	va_end(args);
+}
+
+void mt7996_err(struct mt7996_dev *dev, const char *fmt, ...)
+{
+	struct va_format vaf = {
+		.fmt = fmt,
+	};
+	va_list args;
+
+	va_start(args, fmt);
+	vaf.va = &args;
+	dev_err(dev->mt76.dev, "%pV", &vaf);
+
+	va_end(args);
+}
+
+void mt7996_warn(struct mt7996_dev *dev, const char *fmt, ...)
+{
+	struct va_format vaf = {
+		.fmt = fmt,
+	};
+	va_list args;
+
+	va_start(args, fmt);
+	vaf.va = &args;
+	dev_warn_ratelimited(dev->mt76.dev, "%pV", &vaf);
+
+	va_end(args);
+}
+
+void __mt7996_dbg(struct mt7996_dev *dev, const char *fmt, ...)
+{
+	struct va_format vaf = {
+		.fmt = fmt,
+	};
+	va_list args;
+
+	va_start(args, fmt);
+	vaf.va = &args;
+	dev_printk(KERN_DEBUG, dev->mt76.dev, "%pV", &vaf);
+
+	va_end(args);
+}
+
diff --git a/mt7996/debug.h b/mt7996/debug.h
new file mode 100644
index 00000000..c9454567
--- /dev/null
+++ b/mt7996/debug.h
@@ -0,0 +1,41 @@
+/* SPDX-License-Identifier: ISC */
+/*
+ * Copyright (C) 2024 MediaTek Inc.
+ */
+#ifndef __MT7996_DEBUG_H_
+#define __MT7996_DEBUG_H_
+
+#include "mt7996.h"
+
+enum mt7996_debug_mask {
+	MT7996_DBG_DEV = BIT(0),
+	MT7996_DBG_BSS = BIT(1),
+	MT7996_DBG_STA = BIT(2),
+	MT7996_DBG_CHAN = BIT(3),
+	MT7996_DBG_MLD = BIT(4),
+	MT7996_DBG_TXRX = BIT(5),
+
+	MT7996_DBG_ALL = 0xffffffff,
+};
+
+__printf(2, 3) void mt7996_info(struct mt7996_dev *dev, const char *fmt, ...);
+__printf(2, 3) void mt7996_err(struct mt7996_dev *dev, const char *fmt, ...);
+__printf(2, 3) void mt7996_warn(struct mt7996_dev *dev, const char *fmt, ...);
+
+extern unsigned int mt7996_debug_mask;
+
+__printf(2, 3) void __mt7996_dbg(struct mt7996_dev *dev,
+				 const char *fmt, ...);
+void mt7996_dbg_dump(struct mt7996_dev *dev,
+		     enum mt7996_debug_mask mask,
+		     const char *msg, const char *prefix,
+		     const void *buf, size_t len);
+
+#define mt7996_dbg(dev, dbg_mask, fmt, ...)			\
+do {								\
+	typeof(dbg_mask) mask = (dbg_mask);			\
+	if (mt7996_debug_mask & mask)				\
+		__mt7996_dbg(dev, fmt, ##__VA_ARGS__);	\
+} while (0)
+
+#endif /* __MT7996_DEBUG_H_ */
diff --git a/mt7996/mac.c b/mt7996/mac.c
index dc9ddf82..68cb9b76 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3016,9 +3016,6 @@ void mt7996_scan_work(struct work_struct *work)
 		return;
 	}
 
-	wiphy_info(hw->wiphy, "hw scan %d MHz\n",
-		   req->channels[phy->scan_chan_idx]->center_freq);
-
 	phy->scan_chan = req->channels[phy->scan_chan_idx++];
 
 	if (!req->n_ssids ||
diff --git a/mt7996/main.c b/mt7996/main.c
index 1437cce2..d0e54b6e 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -3,9 +3,11 @@
  * Copyright (C) 2022 MediaTek Inc.
  */
 
+#include <linux/module.h>
 #include "mt7996.h"
 #include "mcu.h"
 #include "mac.h"
+#include "debug.h"
 #ifdef CONFIG_MTK_DEBUG
 #include "mtk_mcu.h"
 #endif
@@ -274,6 +276,12 @@ static void mt7996_remove_bss_conf(struct ieee80211_vif *vif,
 
 	phy = mconf->phy;
 	dev = phy->dev;
+
+	mt7996_dbg(dev, MT7996_DBG_BSS,
+		   "%s: band=%u, bss_idx=%u, link_id=%u, wcid=%u\n",
+		   __func__, mconf->phy->mt76->band_idx, mconf->mt76.idx,
+		   mconf->link_id, mlink->wcid.idx);
+
 	mt7996_mcu_add_sta(dev, conf, mconf, NULL, mlink, false, false);
 	mt7996_mcu_add_bss_info(phy, conf, mconf, mlink, false);
 	mt7996_mcu_add_dev_info(phy, conf, mconf, false);
@@ -423,8 +431,10 @@ static int mt7996_add_bss_conf(struct mt7996_phy *phy,
 	rcu_assign_pointer(mvif->link[link_id], mconf);
 	rcu_assign_pointer(mvif->sta.link[link_id], mlink);
 
-	mlo_dbg(phy, "bss_idx=%u, link_id=%u, wcid=%u\n",
-		mconf->mt76.idx, mconf->link_id, mlink->wcid.idx);
+	mt7996_dbg(dev, MT7996_DBG_BSS,
+		   "%s: band=%u, bss_idx=%u, link_id=%u, wcid=%u\n",
+		   __func__, phy->mt76->band_idx, mconf->mt76.idx,
+		   mconf->link_id, mlink->wcid.idx);
 
 	return 0;
 error:
@@ -717,10 +727,14 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 			add = vif->valid_links ?: BIT(0);
 	}
 
-	mlo_dbg(mt7996_hw_phy(hw), "cipher = 0x%x, icv_len = %u, iv_len = %u, hw_key_idx = %u, keyidx = %d, flags = 0x%x, link_id = %d, keylen = %u\n",
-		     key->cipher, key->icv_len, key->iv_len, key->hw_key_idx, key->keyidx, key->flags, key->link_id, key->keylen);
-	// print_hex_dump(KERN_INFO , "", DUMP_PREFIX_OFFSET, 16, 1, key->key, key->keylen, false);
-	mlo_dbg(mt7996_hw_phy(hw), "add=%lx, valid_links=%x, active_links=%x\n", add, vif->valid_links, vif->active_links);
+	if (sta)
+		mt7996_dbg(dev, MT7996_DBG_STA,
+			   "%s: keyidx=%d, link_bitmap=0x%lx (STA %pM)\n",
+			   __func__, key->keyidx, add, sta->addr);
+	else
+		mt7996_dbg(dev, MT7996_DBG_BSS,
+			   "%s: keyidx=%d, link_bitmap=0x%lx\n",
+			   __func__, key->keyidx, add);
 
 	mutex_lock(&dev->mt76.mutex);
 
@@ -1256,7 +1270,10 @@ static int mt7996_add_link_sta(struct mt7996_dev *dev,
 		rcu_assign_pointer(dev->mt76.wcid[idx], &mlink->wcid);
 		mt76_wcid_init(&mlink->wcid);
 
-		mlo_dbg(mconf->phy, "wcid=%u, link_id=%u, link_addr=%pM, pri_link=%u, sec_link=%u\n", mlink->wcid.idx, link_id, link_sta->addr, msta->pri_link, msta->sec_link);
+		mt7996_dbg(dev, MT7996_DBG_STA,
+			   "%s: STA %pM, wcid=%u, link_id=%u (%pM), pri_link=%u, sec_link=%u\n",
+			   __func__, sta->addr, mlink->wcid.idx, link_id,
+			   link_sta->addr, msta->pri_link, msta->sec_link);
 	}
 
 	if (!assoc)
@@ -1295,7 +1312,7 @@ mt7996_mac_sta_remove_links(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 	if (!ieee80211_vif_is_mld(vif) || rem == sta->valid_links)
 		cancel_delayed_work(&mvif->beacon_mon_work);
 
-	mlo_dbg(mt7996_hw_phy(mvif->hw), "rem=%lu\n", rem);
+	mt7996_dbg(dev, MT7996_DBG_STA, "%s: removed_links=0x%lx\n", __func__, rem);
 	for_each_set_bit(link_id, &rem, IEEE80211_MLD_MAX_NUM_LINKS) {
 		struct mt7996_bss_conf *mconf =
 			mconf_dereference_protected(mvif, link_id);
@@ -1322,7 +1339,8 @@ mt7996_mac_sta_add_links(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 	unsigned int link_id;
 	int i, ret;
 
-	mlo_dbg(mt7996_hw_phy(mvif->hw), "add=%lu, assoc=%d\n", add, assoc);
+	mt7996_dbg(dev, MT7996_DBG_STA,
+		   "%s: added_links=0x%lx, assoc=%d\n", __func__, add, assoc);
 	for_each_set_bit(link_id, &add, IEEE80211_MLD_MAX_NUM_LINKS) {
 		struct mt7996_bss_conf *mconf =
 			mconf_dereference_protected(mvif, link_id);
@@ -1331,6 +1349,9 @@ mt7996_mac_sta_add_links(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 		struct ieee80211_link_sta *link_sta =
 			link_sta_dereference_protected(sta, link_id);
 
+		if (!mconf || !conf || !link_sta)
+			continue;
+
 		ret = mt7996_add_link_sta(dev, conf, mconf, link_sta, assoc);
 		if (ret)
 			goto error;
@@ -1505,13 +1526,16 @@ static void mt7996_tx(struct ieee80211_hw *hw,
 					rcu_dereference(sta->link[link_id]);
 
 				if (!link_sta) {
-					mlo_dbg(mt7996_hw_phy(mvif->hw), "request TX on invalid link_id=%u, use primary link (id=%u) instead.\n",
-						      link_id, msta->pri_link);
+					mt7996_dbg(dev, MT7996_DBG_TXRX,
+						   "%s, request TX on invalid link_id=%u, use primary link (id=%u) instead.\n",
+						   __func__, link_id, msta->pri_link);
 					link_id = msta->pri_link;
 					link_sta = rcu_dereference(sta->link[link_id]);
 
 					if (!link_sta) {
-						mlo_dbg(mt7996_hw_phy(mvif->hw), "primary link became invalid, give up the TX\n");
+						mt7996_dbg(dev, MT7996_DBG_TXRX,
+							   "%s, primary link became invalid, give up the TX\n",
+							   __func__);
 						goto unlock;
 					}
 				}
@@ -2567,6 +2591,8 @@ mt7996_hw_scan(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	struct mt7996_bss_conf *mconf;
 	int ret;
 
+	mt7996_dbg(dev, MT7996_DBG_CHAN, "%s: trigger scan on mt76 band %u\n",
+		   __func__, phy->mt76->band_idx);
 	mutex_lock(&phy->dev->mt76.mutex);
 	if (WARN_ON(phy->scan_req || phy->scan_chan)) {
 		mutex_unlock(&phy->dev->mt76.mutex);
@@ -2698,7 +2724,8 @@ mt7996_add_chanctx(struct ieee80211_hw *hw, struct ieee80211_chanctx_conf *conf)
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
 	int ret;
 
-	wiphy_info(hw->wiphy, "%s: add %u\n", __func__, conf->def.chan->hw_value);
+	mt7996_dbg(dev, MT7996_DBG_CHAN, "%s: add %u on mt76 band %d\n",
+		   __func__, conf->def.chan->hw_value, phy->mt76->band_idx);
 	mutex_lock(&phy->dev->mt76.mutex);
 
 	if (ctx->assigned) {
@@ -2736,8 +2763,10 @@ mt7996_remove_chanctx(struct ieee80211_hw *hw, struct ieee80211_chanctx_conf *co
 {
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
 	struct mt7996_phy *phy = ctx->phy;
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 
-	wiphy_info(hw->wiphy, "%s: remove %u\n", __func__, conf->def.chan->hw_value);
+	mt7996_dbg(dev, MT7996_DBG_CHAN, "%s: remove %u\n",
+		   __func__, conf->def.chan->hw_value);
 	cancel_delayed_work_sync(&phy->scan_work);
 	cancel_delayed_work_sync(&phy->mt76->mac_work);
 
@@ -2755,15 +2784,18 @@ mt7996_change_chanctx(struct ieee80211_hw *hw, struct ieee80211_chanctx_conf *co
 		      u32 changed)
 {
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_phy *phy = ctx->phy;
 
-	wiphy_info(hw->wiphy, "%s: change %u, 0x%x\n", __func__, conf->def.chan->hw_value, changed);
 	if (changed & IEEE80211_CHANCTX_CHANGE_WIDTH ||
 	    changed & IEEE80211_CHANCTX_CHANGE_RADAR) {
 		ctx->chandef = conf->def;
 		phy->mt76->radar_enabled = conf->radar_enabled;
 
 		mt7996_set_channel(phy, &ctx->chandef);
+
+		mt7996_dbg(dev, MT7996_DBG_CHAN, "%s: change to %u, 0x%x\n",
+			   __func__, conf->def.chan->hw_value, changed);
 	}
 }
 
@@ -2773,15 +2805,15 @@ mt7996_assign_vif_chanctx(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			  struct ieee80211_chanctx_conf *conf)
 {
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_phy *phy = ctx->phy;
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
 	struct mt7996_bss_conf *mconf;
 	u8 link_id = link_conf->link_id;
 	int ret;
 
-	wiphy_info(hw->wiphy, "Assign VIF (addr: %pM, type: %d, link_id: %d) to channel context: %d MHz\n",
-		    vif->addr, vif->type, link_conf->link_id,
-		    conf->def.chan->center_freq);
+	mt7996_dbg(dev, MT7996_DBG_CHAN, "%s: assign link_id %u to %d MHz\n",
+		   __func__, link_conf->link_id, conf->def.chan->center_freq);
 
 	mutex_lock(&phy->dev->mt76.mutex);
 
@@ -2820,10 +2852,10 @@ mt7996_unassign_vif_chanctx(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 {
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
 	struct mt7996_phy *phy = ctx->phy;
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 
-	wiphy_info(hw->wiphy, "Remove VIF (addr: %pM, type: %d, link_id: %d) from channel context: %d MHz\n",
-		   vif->addr, vif->type, link_conf->link_id,
-		   conf->def.chan->center_freq);
+	mt7996_dbg(dev, MT7996_DBG_CHAN, "%s, remove link %u from %d MHz\n",
+		   __func__, link_conf->link_id, conf->def.chan->center_freq);
 	cancel_delayed_work_sync(&phy->scan_work);
 
 	mutex_lock(&phy->dev->mt76.mutex);
@@ -2852,9 +2884,16 @@ mt7996_switch_vif_chanctx(struct ieee80211_hw *hw,
 		if (vifs[i].old_ctx == vifs[i].new_ctx)
 			continue;
 
-		wiphy_info(hw->wiphy, "%s: old=%d, new=%d\n",
-			   __func__, vifs[i].old_ctx->def.chan->hw_value,
-			   vifs[i].new_ctx->def.chan->hw_value);
+		mt7996_dbg(dev, MT7996_DBG_CHAN,
+			   "%s: chan=%d->%d, width=%d->%d, punct_bitmap=0x%04x->0x%04x, link=%u\n",
+			   __func__,
+			   vifs[i].old_ctx->def.chan->hw_value,
+			   vifs[i].new_ctx->def.chan->hw_value,
+			   vifs[i].old_ctx->def.width,
+			   vifs[i].new_ctx->def.width,
+			   vifs[i].old_ctx->def.punctured,
+			   vifs[i].new_ctx->def.punctured,
+			   vifs[i].link_conf->link_id);
 
 		mutex_lock(&dev->mt76.mutex);
 
@@ -2904,7 +2943,10 @@ mt7996_change_vif_links(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	unsigned int link_id;
 	int ret = 0;
 
-	mlo_dbg(phy, "old=%u, new=%u\n", old_links, new_links);
+	mt7996_dbg(dev, MT7996_DBG_MLD,
+		   "%s: old=0x%x, new=0x%x, dormant=0x%x\n",
+		   __func__, old_links, new_links, vif->dormant_links);
+
 	if (old_links == new_links)
 		return 0;
 
@@ -2956,7 +2998,8 @@ mt7996_change_sta_links(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	unsigned long rem = old_links & ~new_links;
 	int ret = 0;
 
-	mlo_dbg(mt7996_hw_phy(hw), "old=%u, new=%u\n", old_links, new_links);
+	mt7996_dbg(dev, MT7996_DBG_STA, "%s: STA %pM old=0x%x, new=0x%x\n",
+		   __func__, sta->addr, old_links, new_links);
 	mutex_lock(&dev->mt76.mutex);
 
 	if (rem)
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index f7ee27f9..1bfb783f 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -10,6 +10,7 @@
 #include "mac.h"
 #include "eeprom.h"
 #include "mt7996_trace.h"
+#include "debug.h"
 
 #define fw_name(_dev, name, ...)	({			\
 	char *_fw;						\
@@ -386,7 +387,6 @@ mt7996_mcu_csa_finish(void *priv, u8 *mac, struct ieee80211_vif *vif)
 	if (!link_conf || !link_conf->csa_active || vif->type == NL80211_IFTYPE_STATION)
 		return;
 
-	pr_info("%s: link_id=%d\n", __func__, link_id);
 	mvif->cs_ready_links = 0;
 	mvif->cs_link_id = IEEE80211_LINK_UNSPECIFIED;
 	ieee80211_csa_finish(vif, link_id);
@@ -1491,6 +1491,7 @@ mt7996_mcu_bss_mld_tlv(struct sk_buff *skb, struct ieee80211_vif *vif,
 		       struct mt7996_bss_conf *mconf)
 {
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
+	struct mt7996_dev *dev = mconf->phy->dev;
 	struct bss_mld_tlv *mld;
 	struct tlv *tlv;
 
@@ -1507,8 +1508,10 @@ mt7996_mcu_bss_mld_tlv(struct sk_buff *skb, struct ieee80211_vif *vif,
 	}
 
 	mld->own_mld_id = mconf->own_mld_id;
-	pr_info("%s: group_mld_id=%d own_mld_id=%d remap_idx=%d mld->addr[%pM]\n",
-		__func__, mld->group_mld_id,  mld->own_mld_id, mld->remap_idx, mld->mac_addr);
+	mt7996_dbg(dev, MT7996_DBG_BSS,
+		   "%s: group_mld_id=%d, own_mld_id=%d, remap_idx=%d, mld->addr[%pM]\n",
+		   __func__, mld->group_mld_id,  mld->own_mld_id,
+		   mld->remap_idx, mld->mac_addr);
 }
 
 static void
@@ -1595,6 +1598,7 @@ mt7996_mcu_bss_basic_tlv(struct sk_buff *skb, struct ieee80211_bss_conf *conf,
 {
 	struct mt76_phy *phy = mconf->phy->mt76;
 	struct ieee80211_vif *vif = conf->vif;
+	struct mt7996_dev *dev = mconf->phy->dev;
 	struct cfg80211_chan_def *chandef = &phy->chandef;
 	struct mt76_connac_bss_basic_tlv *bss;
 	u32 type = CONNECTION_INFRA_AP;
@@ -1661,8 +1665,10 @@ mt7996_mcu_bss_basic_tlv(struct sk_buff *skb, struct ieee80211_bss_conf *conf,
 
 	memcpy(bss->bssid, conf->bssid, ETH_ALEN);
 
-	mlo_dbg(mconf->phy, "omac_idx=%d band_idx=%d wmm_idx=%d bss->bssid=%pM enable=%d\n",
-		bss->omac_idx, bss->band_idx, bss->wmm_idx, bss->bssid, enable);
+	mt7996_dbg(dev, MT7996_DBG_BSS,
+		   "%s: band=%d, omac=%d, wmm_idx=%d, bssid=%pM, link=%d, en=%d\n",
+		   __func__, bss->band_idx, bss->omac_idx,
+		   bss->wmm_idx, bss->bssid, conf->link_id, enable);
 
 	bss->bcn_interval = cpu_to_le16(conf->beacon_int);
 	bss->dtim_period = conf->dtim_period;
@@ -1801,8 +1807,6 @@ mt7996_mcu_sta_tx_cap(struct mt7996_dev *dev, struct mt76_vif *mvif,
 	tx_cap = (struct sta_rec_tx_cap *)tlv;
 	tx_cap->ampdu_limit_en = true;
 
-	dev_info(dev->mt76.dev, "%s: limit wcid %d ampdu to 512\n", __func__, wcid->idx);
-
 	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
 				     MCU_WMWA_UNI_CMD(STA_REC_UPDATE), true);
 }
@@ -2978,8 +2982,10 @@ int mt7996_mcu_add_sta(struct mt7996_dev *dev, struct ieee80211_bss_conf *conf,
 	conn_state = enable ? CONN_STATE_PORT_SECURE : CONN_STATE_DISCONNECT;
 	mt76_connac_mcu_sta_basic_tlv(&dev->mt76, skb, conf, link_sta,
 				      conn_state, newly);
-	mlo_dbg(mconf->phy, "link=%u, newly=%d, conn_state=%d\n",
-		mlink->wcid.link_id, newly, conn_state);
+	mt7996_dbg(dev, MT7996_DBG_DEV,
+		   "%s: link=%u, wcid=%u, newly=%d, conn_state=%d\n",
+		   __func__, mlink->wcid.link_id, mlink->wcid.idx,
+		   newly, conn_state);
 
 	if (!enable)
 		goto out;
@@ -3039,6 +3045,8 @@ mt7996_mcu_sta_mld_setup_tlv(struct mt7996_dev *dev, struct sk_buff *skb,
 	struct tlv *tlv;
 	unsigned long valid_links = sta->valid_links;
 	unsigned int link_id;
+	struct ieee80211_vif *vif = container_of((void *)msta->vif, struct ieee80211_vif,
+						 drv_priv);
 
 	mlink = mlink_dereference_protected(msta, msta->pri_link);
 	if (!mlink)
@@ -3062,15 +3070,16 @@ mt7996_mcu_sta_mld_setup_tlv(struct mt7996_dev *dev, struct sk_buff *skb,
 	mld_setup->link_num = hweight16(sta->valid_links);
 
 	mld_setup_link = (struct mld_setup_link *)mld_setup->link_info;
-	mlo_dbg(mt7996_hw_phy(mlink->sta->vif->hw), "pri_link(%u) primary_id(%d) seconed_id(%d) wcid(%d), link_num(%d), mld_addr[%pM]\n",
-		msta->pri_link, mld_setup->primary_id, mld_setup->seconed_id, mld_setup->setup_wcid, mld_setup->link_num, mld_setup->mld_addr);
+	mt76_trace(vif, "STA %pM pri_link=%u, pri_wcid=%u, sec_link=%u, sec_wcid=%u\n",
+		   sta->addr, msta->pri_link, le16_to_cpu(mld_setup->primary_id),
+		   msta->sec_link, le16_to_cpu(mld_setup->seconed_id));
 	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
 		mlink = mlink_dereference_protected(msta, link_id);
 		mconf = mconf_dereference_protected(msta->vif, link_id);
 
 		mld_setup_link->wcid = cpu_to_le16(mlink->wcid.idx);
 		mld_setup_link->bss_idx = mconf->mt76.idx;
-		mlo_dbg(mt7996_hw_phy(mlink->sta->vif->hw), "link_id(%d) wcid(%d) bss_idx(%d)\n",
+		mt76_trace(vif, "link_id(%d) wcid(%d) bss_idx(%d)\n",
 		link_id, mld_setup_link->wcid, mld_setup_link->bss_idx);
 		mld_setup_link++;
 	}
@@ -3336,8 +3345,9 @@ int mt7996_mcu_add_dev_info(struct mt7996_phy *phy,
 		return mt7996_mcu_muar_config(phy, conf, mconf, false, enable);
 
 	memcpy(data.tlv.omac_addr, conf->addr, ETH_ALEN);
-	mlo_dbg(phy, "omac=%u, band=%u, addr=%pM, en=%d\n",
-		data.hdr.omac_idx,data.hdr.band_idx, data.tlv.omac_addr, enable);
+	mt7996_dbg(dev, MT7996_DBG_DEV, "%s: band=%u, omac=%u, addr=%pM, en=%d\n",
+		   __func__, data.hdr.band_idx, data.hdr.omac_idx,
+		   data.tlv.omac_addr, enable);
 	return mt76_mcu_send_msg(&dev->mt76, MCU_WMWA_UNI_CMD(DEV_INFO_UPDATE),
 				 &data, sizeof(data), true);
 }
@@ -6716,10 +6726,11 @@ int mt7996_mcu_set_eml_omn(struct ieee80211_hw *hw,
 		}
 	}
 
-	mlo_dbg(mconf->phy, "link:%u, wcid:%d, control:%x, mode:%d, bmp:%x\n",
-		mlink->wcid.link_id, mlink->wcid.idx, eml_omn->control,
-		!!(eml_omn->control & EML_OMN_CONTROL_EMLSR_MODE),
-		eml_op->bitmap);
+	mt7996_dbg(dev, MT7996_DBG_MLD,
+		   "%s: link:%u, wcid:%d, control:%x, mode:%d, bmp:%x\n",
+		   __func__, mlink->wcid.link_id, mlink->wcid.idx, eml_omn->control,
+		   !!(eml_omn->control & EML_OMN_CONTROL_EMLSR_MODE),
+		   eml_op->bitmap);
 
 	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
 			MCU_WMWA_UNI_CMD(STA_REC_UPDATE), true);
-- 
2.45.2

